More rules logging and L003 indent improvement.

alanmcruickshank opened the pull request:
This resolves #734 (or at least part of it).

This fix for the original query is now:

```sql
SELECT
    *
FROM foo
INNER JOIN bar
    USING (baz)
WHERE qux = 1
```

however, if the `bar` is on a new line it still fixes to:

```sql
SELECT
    *
FROM foo
INNER JOIN
        bar
    USING (baz)
WHERE qux = 1
```

I've also tried to add a load more logging in. @barrywhart - does that also make the structure easier to understand?

alanmcruickshank commented:
We might want to revisit the latter problem once #882 is merged, because I think that might make things easier - and might also fix it.

alanmcruickshank commented:
BOOM 🎉 

Now that #882 is merged it also fixes the indentation of `bar`:

```sql
SELECT
    *
FROM foo
INNER JOIN
    bar
    USING (baz)
WHERE qux = 1
```

barrywhart reviewed (commented):
The code changes look good. A couple of existing tests broke, and maybe we want to add a test or two for the new behavior?

barrywhart commented:
Some thoughts on other things that could make L003 easier to understand and modify:
* The concept of `clean_indent` seems pretty important to the design, but is hard to grasp from the existing comments. It might be helpful to add test cases to `_process_raw_stack()`, potentially using pytest "[parametrize](https://docs.pytest.org/en/stable/parametrize.html)", to explain by example
* Maybe the same as above but for detecting "hanging indent"
* It could be helpful to create smaller tests for the same behavior as the two large tests, but finer grained
  * `test/core/rules/std_fix_auto_test.py:143 test__std_fix_auto[ansi-002_indentation]`
  * `test/core/rules/std_fix_auto_test.py:143 test__std_fix_auto[bigquery-004_templating]`

alanmcruickshank commented:
@barrywhart - can I get a re-review? I've added a load of test cases and commenting to improve coverage.

Fixed a few bugs too in the process. 🚀 

The `bigquery` templating example is still behaving a little strange, but I think that's more to do with the templating than L003.

barrywhart commented on test/core/rules/std_L003_test.py:
```diff
@@ -0,0 +1,173 @@
+@pytest.mark.parametrize(
+    "test_elems,result",
+    [
+        # Unindented test examples
```
I 💕 these detailed test cases!

barrywhart commented on test/core/rules/test_cases/L003.yml:
```diff
@@ -1,24 +1,26 @@
         acceptable_buckets
 
-test_4:
+test__identation_of_comments_2:
```
Typos:
* Two underscores after "test"
* identation -> indentation

barrywhart commented on test/core/rules/test_cases/L003.yml:
```diff
@@ -167,3 +165,45 @@ bigquery_with_indent:
+    FROM tbl
+
+test_possible_hanger_fix:
+  # It's not even close to a hnaging indent, but we should
```
Typo: hnaging -> hanging

barrywhart commented on test/core/rules/test_cases/L003.yml:
```diff
@@ -1,24 +1,26 @@
 rule: L003
 
-test_1:
```
Great tests! Love the new ones, the names, and the comments. My 🧠 is so happy now, ha ha.

barrywhart commented on test/fixtures/linter/autofix/bigquery/004_templating/after.sql:
```diff
@@ -24,6 +24,8 @@ raw_effect_sizes AS (
     SELECT
         COUNT(1) AS campaign_count_{{action}},
         {{corr_states}}
+    -- NOTE: The L003 fix routine behaves a little strangely here around the templated
```
Can you clarify what is the strange behavior seen here? Is it the indentation of the `STDDEV_POP(` line?

barrywhart commented on test/fixtures/linter/autofix/bigquery/004_templating/before.sql:
```diff
@@ -24,6 +24,8 @@ WITH
   SELECT
     COUNT(1) AS campaign_count_{{action}},
     {{corr_states}}
+    -- NOTE: The L003 fix routine behaves a little strangely here around the templated
```
Also here.

barrywhart reviewed (approved):
This is great! 🎉💯

I found a few typos, and there are a couple of places where comments could be clearer. Feel free to merge once you're ready; I don't see a need for any additional review.
